Mode select

AdamHaffar opened the pull request:
Description 

- Updated to work with new `SDK` and `Conduit` Versions.
- Adds 2 modes (Snapshot & CDC) to the Airtable Connector using a bool for `EnableCDC` parameter.
- Snapshot mode takes a snapshot of the current Airtable records but does not detect changes.
- CDC mode can detect `snapshot`, `created` and `updated` records from airtable. This requires some manual configuration of Airtable records specified in the `ReadMe` to work.

AdamHaffar committed:
add cdc bool to config struct and parse it

AdamHaffar committed:
add cdc and combined iterator files

AdamHaffar committed:
remove combined and cdc iterator and update HasNext

AdamHaffar committed:
update iterator to fetch in pages

AdamHaffar committed:
implement CDC iterator to create if the CDC boolean is true

AdamHaffar committed:
implement cdc_iterator.go

AdamHaffar committed:
fix dependency path

AdamHaffar committed:
fix buildRecordKey to display correct information

AdamHaffar committed:
change cdc_iterator to detect updates in airtable using 2 fields

AdamHaffar committed:
-add a nil check for position when creating newCDCiterator so position is not lost.
-initial value of s.position.RecordSlicePos set to -1 to fix indexing errors.
-Use CreatedTime variable to compare again s.position.LastKnownTime to determine RecordAction (CREATED,UPDATED,SNAPSHOT) and implement each respective sdk.Util.Source.NewRecord
-commented code

hariso commented on config/config.go:
```diff
@@ -57,11 +60,16 @@ func ParseBaseConfig(cfg map[string]string) (Config, error) {
 		return Config{}, err
 	}
 
+	CDCval, err := strconv.ParseBool(cfg[EnableCDC]) //Convert string value
```
Local variables, such as `CDCval` should be lower-case.

hariso commented on iterator/snapshot_iterator.go:
```diff
@@ -3,26 +3,33 @@ package iterator
 	logger.Trace().Msg("Creating new snapshot iterator")
 
 	table := client.GetTable(config.BaseID, config.TableID)
+	IteratorConfig := config
```
Why is this assignment needed?

AdamHaffar replied:
not needed

hariso commented on iterator/snapshot_iterator.go:
```diff
@@ -3,26 +3,33 @@ package iterator
+	IteratorConfig := config
+
 	records, err := table.GetRecords().
 		InStringFormat("Europe/London", "en-gb").
```
Hard-coded values tend to cause issues down the road. Is there a reason why we need to hard-code this time zone?

AdamHaffar replied:
not necessary, removed

AdamHaffar replied:
not needed, removed

hariso commented on iterator/snapshot_iterator.go:
```diff
@@ -3,26 +3,33 @@ package iterator
 		Do()
+
 	if err != nil {
 		return &SnapshotIterator{}, fmt.Errorf("error while getting records")
```
I'd suggest changing the first return to `nil`. There's no need to return anything initialized because we have an error.

Also, it's good practice to wrap errors in Go, because the original error carries important information. You can do that with `fmt.Errorf("error while getting records: %w", err)`. This will create a new error, which wraps `err`.

hariso commented on iterator/snapshot_iterator.go:
```diff
@@ -3,26 +3,33 @@ package iterator
 	table := client.GetTable(config.BaseID, config.TableID)
+	IteratorConfig := config
+
 	records, err := table.GetRecords().
```
Why do we need to get records when creating an iterator?

AdamHaffar replied:
not needed, removed

hariso commented on iterator/snapshot_iterator.go:
```diff
@@ -31,23 +38,32 @@ func NewSnapshotIterator(ctx context.Context, client *airtableclient.Client, con
 
-	if s.internalPos.Index == len(s.data.Records) {
-		return false
+		if s.position.Offset == "" { //Checks if last page (no offset)
```
Nice comments!

hariso commented on iterator/snapshot_iterator.go:
```diff
@@ -61,9 +77,13 @@ func (s *SnapshotIterator) Next(ctx context.Context) (sdk.Record, error) {
-	s.internalPos.Index++ // increment internal position
+	pos, err := s.buildRecordPosition()
+	if err != nil {
+		return sdk.Record{}, err
```
It would be good to wrap the error and add some context around it, e.g. `fmt.Errorf("failed building position: %w", err)`.

hariso commented on iterator/snapshot_iterator.go:
```diff
@@ -61,9 +77,13 @@ func (s *SnapshotIterator) Next(ctx context.Context) (sdk.Record, error) {
 	}
 
-	s.internalPos.Index++ // increment internal position
+	pos, err := s.buildRecordPosition()
```
In `s.buildPosition()` we have this:
```go
s.position.RecordSlicePos++ // increment internal position
```
IMHO it would be better if you do move the position outside the method. Otherwise, `s.buildPosition()` always has to be called first. If you would do this:
```go
rec := sdk.Util.Source.NewRecordSnapshot(
    nil,
    s.buildRecordMetadata(),
    s.buildRecordKey(),
    s.buildRecordPayload(),
 )
 rec.Position = pos
 ```
 you would get different results, which would be unexpected.

hariso commented on iterator/snapshot_iterator.go:
```diff
@@ -3,26 +3,33 @@ package iterator
-	internalPos position.Position
+	client   *airtableclient.Client
+	data     *airtableclient.Records
+	position position.Position
```
I believe it would help if we rename this field to `currentPosition`.

hariso commented on source/source.go:
```diff
@@ -50,10 +49,20 @@ func (s *Source) Open(ctx context.Context, pos sdk.Position) error {
-	if err != nil {
-		logger.Error().Stack().Err(err).Msg("Error while creating iterator")
-		return fmt.Errorf("couldn't create an iterator: %w", err)
+	if s.config.EnableCDC {
+		s.iterator, err = iterator.NewCDCIterator(ctx, s.client, s.config, pos)
+		if err != nil {
+			logger.Error().Stack().Err(err).Msg("Error while creating cdc iterator")
+			return fmt.Errorf("couldn't create a cdc iterator: %w", err)
+		}
+	}
+
+	if !s.config.EnableCDC {
```
These two conditions naturally fit into a `if-then-else`, so I'd suggest that.

hariso commented on source/source.go:
```diff
@@ -50,10 +49,20 @@ func (s *Source) Open(ctx context.Context, pos sdk.Position) error {
+	if !s.config.EnableCDC {
+		s.iterator, err = iterator.NewSnapshotIterator(ctx, s.client, s.config, pos)
+		if err != nil {
+			logger.Error().Stack().Err(err).Msg("Error while creating snapshot iterator")
```
Nice touch with `Stack()`!

hariso commented on iterator/snapshot_iterator.go:
```diff
@@ -3,26 +3,33 @@ package iterator
-	data        *airtableclient.Records
-	internalPos position.Position
+	client   *airtableclient.Client
+	data     *airtableclient.Records
```
`data` might be a bit generic in this context... What do you think about `records` or `currentPage`?

hariso commented on go.mod:
```diff
@@ -1,4 +1,4 @@
+module github.com/conduitio-labs/conduit-connector-airtable
 
 go 1.18
 
```
SDK 0.5 is still hot off the press.:)

AdamHaffar committed:
readme

AdamHaffar committed:
readme

hariso committed:
added Makefile, update gitignore

hariso committed:
added Makefile, update gitignore

AdamHaffar committed:
resolve issues from pull request suggested by @Haris:

-wrap errors
-remove unnecessary config assignment
-remove hardcoded values
-change increment order to outside of buildRecordPosition()

AdamHaffar committed:
Merge remote-tracking branch 'origin/mode-select' into mode-select

AdamHaffar committed:
remove hardcoded field values and add consts for them

AdamHaffar committed:
remove hardcoded field values and add consts for them

hariso commented on iterator/cdc_iterator.go:
```diff
@@ -0,0 +1,212 @@
+	"time"
+)
+
+const LASTMODIFIED = "last-modified"
+const LASTMODIFIEDSTR = "last-modified-str"
```
It's great that you extracted this into constants! A nitpick related to this: it's a convention to use mixed caps for naming constants (in Go at least): https://github.com/golang/go/wiki/CodeReviewComments#mixed-caps.

hariso commented on config/config.go:
```diff
@@ -57,11 +60,16 @@ func ParseBaseConfig(cfg map[string]string) (Config, error) {
 
+	cdcval, err := strconv.ParseBool(cfg[EnableCDC]) //Convert string value
+	if !ok {
+		return Config{}, fmt.Errorf("%v", err)
```
`fmt.Errorf("%v", err)` will kinda copy `err` into a new one. We generally do one of things when returning errors:
1. Either return them directly, e.g. `return err`
2. Or we add some context and then return it `fmt.Errorf("cannot parse EnableCDC, input value %v: %w", cdcVal, err)`.

AdamHaffar committed:
resolve issues and update to sdk 6.0

AdamHaffar committed:
Merge remote-tracking branch 'origin/mode-select' into mode-select

# Conflicts:
#	iterator/cdc_iterator.go

AdamHaffar committed:
add config unit tests and update config.go check if EnableCDC is empty
